src/ZkFold/Base/Protocol/Protostar/Fold.hs

```diff
@@ -0,0 +1,179 @@
+import           Prelude                                             (and, otherwise, type (~), ($), (<$>), (<*>), (==))
+import qualified Prelude                                             as P
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Number
+import qualified ZkFold.Base.Algebra.Polynomials.Univariate          as PU
+import           ZkFold.Base.Data.Vector                             (Vector)
+import           ZkFold.Base.Protocol.Protostar.Accumulator
+import qualified ZkFold.Base.Protocol.Protostar.AccumulatorScheme    as Acc
+import           ZkFold.Base.Protocol.Protostar.ArithmeticCircuit
+import           ZkFold.Base.Protocol.Protostar.Commit
+import           ZkFold.Base.Protocol.Protostar.CommitOpen
+import           ZkFold.Base.Protocol.Protostar.FiatShamir
+import           ZkFold.Base.Protocol.Protostar.Oracle
+import qualified ZkFold.Base.Protocol.Protostar.SpecialSound         as SPS
+import           ZkFold.Prelude                                      (replicate)
+import           ZkFold.Symbolic.Compiler
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Data.FieldElement                   (FieldElement)
+
+data FoldResult n c a
+    = FoldResult
+    { output          :: Vector n a
+    , lastAccumulator :: Accumulator (Vector n a) a c (Map Natural a)
+    , verifierOutput  :: P.Bool
+    , deciderOutput   :: P.Bool
+    } deriving (P.Show, Generic, NFData)
+
+type FS_CM n c a = FiatShamir a (CommitOpen a c (RecursiveCircuit n a))
+
+transform
+    :: forall n c a
+    .  HomomorphicCommit a [Map Natural a] c
+    => a
+    -> RecursiveCircuit n a
+    -> Vector n a
+    -> FS_CM n c a
+transform ck rc v = FiatShamir (CommitOpen (hcommit ck) rc) v
+
+-- | These instances might seem off, but accumulator scheme requires this exact behaviour
+--
+instance AdditiveSemigroup a => AdditiveSemigroup (Map Natural a) where
+    (+) = M.unionWith (+)
+
+instance AdditiveMonoid a => AdditiveMonoid (Map Natural a) where
+    zero = M.empty
+
+instance (Ring a, P.Eq a) => Acc.LinearCombination (Map Natural a) (Map Natural (PU.Poly a)) where
+    linearCombination mx ma = M.unionWith (+) (PU.monomial 1 <$> mx) (PU.constant <$> ma)
+
+instance (Ring a, P.Eq a, KnownNat n) => Acc.LinearCombination (Vector n a) (Vector n (PU.Poly a)) where
+    linearCombination mx ma = (+) <$> (PU.monomial 1 <$> mx) <*> (PU.constant <$> ma)
+
+instance (Ring a, KnownNat n) => Acc.LinearCombinationWith a (Vector n a) where
+    linearCombinationWith coeff a b = (+) <$> (P.fmap (coeff *) a) <*> b
+
+fold
+    :: forall a n c x
+    .  Arithmetic a
+    => x ~ ArithmeticCircuit a (Vector n)
+    => P.Eq c
+    => Scale a c
+    => Scale a a
+    => AdditiveGroup c
+    => RandomOracle a a
+    => RandomOracle c a
+    => RandomOracle [c] a
+    => HomomorphicCommit a [a] c
+    => HomomorphicCommit a [Map Natural a] c
```

Comment:
These look like references to concrete instances, maybe it is better to inline them then?